### PR TITLE
Trie regression

### PR DESCRIPTION
**Pull request by msune**
Fix bug during flowmod insertion in trie.

**WARNING**: this PR depends/contains/is based on #11

### CODE REVIEW

**Comment by msune** [comment]
It would be advisable to run the entire test suite of RYU to see if it has any regression. I doubt it but you never know.

Note that RYU test-cases weren't hitting this assertion.

**Comment by msune** [comment]
Fixed an issue with the test with `libcunit`. Some versions do not accept tests to have the same description string

**Comment by toanju** [comment]
tests run only in a build folder, that is one level deeper in the directory tree than the source (like the provided build folder). Building anywhere else fails:

```
gcc -DPACKAGE_NAME=\"ROFL-DATAPATH\" -DPACKAGE_TARNAME=\"rofl-datapath\" -DPACKAGE_VERSION=\"v0.6.4\" -DPACKAGE_STRING=\"ROFL-DATAPATH\ v0.6.4\" -DPACKAGE_BUGREPORT=\"rofl-devel@roflibs.org\" -DPACKAGE_URL=\"http://www.roflibs.org\" -DPACKAGE=\"rofl-datapath\" -DVERSION=\"v0.6.4\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -D__STDC_FORMAT_MACROS=/\*\*/ -DDEBUG=/\*\*/ -DHAVE_MA_TRIE=1 -DHAVE_MA_LOOP=1 -DHAVE_MA_L2HASH=1 -I. -I/home/tobi/workspace/rofl-datapath/src/ -I/home/tobi/workspace/rofl-datapath/src/ -I../src/ -I/home/tobi/workspace/rofl-datapath/test/rofl/datapath/pipeline/../src/ -I/home/tobi/workspace/rofl-datapath/test/rofl/datapath/pipeline/../src/ -I/home/tobi/workspace/rofl-datapath/test/rofl/datapath/pipeline/../src/ -I/home/tobi/workspace/rofl-datapath/test/rofl/datapath/pipeline/../src/ -DROFL_TEST=1   -std=gnu89 -Werror -Wall  -fno-omit-frame-pointer -g -g -O0 -D__COMPILING_ROFL_PIPELINE__ -c -o openflow/openflow1x/pipeline/memory.o openflow/openflow1x/pipeline/memory.c
tern_alike_test.c:6:77: fatal error: ../../../../../src/rofl/datapath/pipeline/common/ternary_fields.h: No such file or directory
 #include "../../../../../src/rofl/datapath/pipeline/common/ternary_fields.h"
                                                                             ^
```

**Comment by toanju** [comment]
The code could be rebased now to integration-0.7, since #11 is merged

**Comment by msune** [comment]
Good catch. Compilation error fixed. Actually, it was introduced by #8.

Rebase done. It would have been easier to just integrate #12, but that's fine :smile:

File: test/rofl/datapath/pipeline/openflow/openflow1x/pipeline/ma/trie/trie.c [hunk 1]
```diff
@@ -958,6 +958,57 @@ void test_regressions(){
...
+	CU_ASSERT(of1x_add_flow_entry_table(&sw->pipeline, 0, &entry, false,false) == ROFL_OF1X_FM_SUCCESS);
+
+	of1x_full_dump_switch(sw, false);
+	CU_ASSERT(table->num_of_entries == 1);
+
+	//Add one with the same + 1 match
+	entry = of1x_init_flow_entry(false);
+	CU_ASSERT(entry != NULL);
+
+#if 0
```
Comment (toanju) [inline test/rofl/datapath/pipeline/openflow/openflow1x/pipeline/ma/trie/trie.c]: dead code added

**Comment by toanju** [comment]
Interesting, because #11 was compiling fine in a different folder. On the other hand it's still not working:

```
gcc -DPACKAGE_NAME=\"ROFL-DATAPATH\" -DPACKAGE_TARNAME=\"rofl-datapath\" -DPACKAGE_VERSION=\"v0.6.4\" -DPACKAGE_STRING=\"ROFL-DATAPATH\ v0.6.4\" -DPACKAGE_BUGREPORT=\"rofl-devel@roflibs.org\" -DPACKAGE_URL=\"http://www.roflibs.org\" -DPACKAGE=\"rofl-datapath\" -DVERSION=\"v0.6.4\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -D__STDC_FORMAT_MACROS=/\*\*/ -DDEBUG=/\*\*/ -DHAVE_MA_TRIE=1 -DHAVE_MA_LOOP=1 -DHAVE_MA_L2HASH=1 -I. -I/home/tobi/workspace/rofl-datapath/src/ -I/home/tobi/workspace/rofl-datapath/src/ -I../src/ -I/home/tobi/workspace/rofl-datapath/test/rofl/datapath/pipeline/../src/ -I/home/tobi/workspace/rofl-datapath/test/rofl/datapath/pipeline/../src/ -I/home/tobi/workspace/rofl-datapath/test/rofl/datapath/pipeline/../src/ -I/home/tobi/workspace/rofl-datapath/test/rofl/datapath/pipeline/../src/ -DROFL_TEST=1   -std=gnu89 -Werror -Wall  -fno-omit-frame-pointer -g -g -O0 -D__COMPILING_ROFL_PIPELINE__ -c -o pipeline_/common/alike_masks.o pipeline_/common/alike_masks.c
tern_alike_test.c:6:77: fatal error: ../../../../../src/rofl/datapath/pipeline/common/ternary_fields.h: No such file or directory
 #include "../../../../../src/rofl/datapath/pipeline/common/ternary_fields.h"
```

Be aware, that this is one of the tests.

Rebasing just gives a nicer history, so I prefer that.

File: test/rofl/datapath/pipeline/tern_alike_test.c [hunk 2]
```diff
@@ -3,7 +3,7 @@
 #include <assert.h>
 #include <arpa/inet.h>
 #include "CUnit/Basic.h"
-#include "../../../../../src/rofl/datapath/pipeline/common/ternary_fields.h"
+#include "rofl/datapath/pipeline/common/ternary_fields.h"
```
Comment (toanju) [inline test/rofl/datapath/pipeline/tern_alike_test.c:6]: probably the right include is missing for this one

File: test/rofl/datapath/pipeline/openflow/openflow1x/pipeline/ma/trie/trie.c (hunk 1 above)
Comment (msune) [reply test/rofl/datapath/pipeline/openflow/openflow1x/pipeline/ma/trie/trie.c]: I can remove it, it was from PR #11

**Comment by msune** [comment]
It could not compile before. PR #12 does not change anything,

**Comment by msune** [comment]
Do  a `maintainer-clean` for testing. It works fine here.

File: test/rofl/datapath/pipeline/tern_alike_test.c (hunk 2 above)
Comment (msune) [reply test/rofl/datapath/pipeline/tern_alike_test.c:6]: This is the fix. I don't understand your comment.

**Comment by msune** [comment]
>>Be aware, that this is one of the tests.
>>Rebasing just gives a nicer history, so I prefer that.

On this; since Pr #12 included PR #11, that is, it was rebased to the PR #11 base, it would have been easier to just integrate PR #12, without letting github introduce the fake commit (because it was fast-forward) of merge that github UI does, hence PR #12 would have been FF. That's what I meant.

File: test/rofl/datapath/pipeline/openflow/openflow1x/pipeline/ma/trie/trie.c (hunk 1 above)
Comment (msune) [reply test/rofl/datapath/pipeline/openflow/openflow1x/pipeline/ma/trie/trie.c]: Fixed

Comment (toanju) [reply test/rofl/datapath/pipeline/openflow/openflow1x/pipeline/ma/trie/trie.c]: indeed, I overlooked it in #11, and github still showed it in the non rebased version... Thanks for removing

**Comment by toanju** [comment]
sorry, something went south in my setup... somehow I wasn't able to force pull the updates and that's were I got wrong here

**Pull request merged**